tasks.go

```diff
@@ -1,125 +1,162 @@
+		wg.Wait()
+		close(results)
+	}()
+	// Loop through the task execution results and log errors.
+	for err := range results {
+		if err != nil {
+			// TODO: there should be a way to identify which task
+			// had an error.
+			fmt.Fprintln(os.Stderr)
+			log.Printf("Task error: %v", err)
+			continue
+		}
+		fmt.Fprint(os.Stderr, ".")
+	}
+	fmt.Fprintln(os.Stderr)
+}
 
+// GetAllTasks returns a channel of all tasks known to the dump tool. It returns
+// immediately and sends tasks to the channel in a separate goroutine. The
+// channel is closed after all tasks are sent.
+// FIXME: GetAllTasks should not need to know about basepath.
+func GetAllTasks(basepath string) <-chan Task {
 	var (
 		resources = []string{"deploymentconfigs", "pods", "services", "events"}
 		// We should only care about logs for pods, because they cover
 		// all other possible types.
 		resourcesWithLogs = []string{"pods"}
 	)
+	tasks := make(chan Task)
+	go func() {
+		defer close(tasks)
 
-	projects, err := GetProjects()
-	if err != nil {
-		return nil, err
-	}
-
-	if len(projects) == 0 {
-		return nil, errors.New("no projects visible to the currently logged in user")
-	}
-
-	// Add tasks to fetch resource definitions.
-	definitionsTasks, err := GetResourceDefinitionsTasks(projects, resources, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, definitionsTasks...)
-
-	// Add tasks to fetch logs.
-	logsTasks, err := GetFetchLogsTasks(projects, resourcesWithLogs, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, logsTasks...)
-
-	nagiosDataTasks, err := GetNagiosTasks(projects, basepath)
-	if err != nil {
-		retErrors = append(retErrors, err)
-	}
-	tasks = append(tasks, nagiosDataTasks...)
+		projects, err := GetProjects()
+		if err != nil {
+			tasks <- NewError(err)
+			return
+		}
+		if len(projects) == 0 {
+			tasks <- NewError(errors.New("no projects visible to the currently logged in user"))
+			return
+		}
 
-	// Add check tasks
-	for _, p := range projects {
-		outFor := outToFile(basepath, "json", "analysis")
-		errOutFor := outToFile(basepath, "stderr", "analysis")
-		task := CheckTasks(p, outFor, errOutFor)
-		tasks = append(tasks, task)
-	}
+		var wg sync.WaitGroup
+
+		// Add tasks to fetch resource definitions.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			GetResourceDefinitionsTasks(tasks, projects, resources, basepath)
+		}()
+
+		// Add tasks to fetch logs.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			GetFetchLogsTasks(tasks, projects, resourcesWithLogs, basepath)
+		}()
+
+		// Add tasks to fetch Nagios data.
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			GetNagiosTasks(tasks, projects, basepath)
+		}()
+
+		wg.Wait()
+
+		// After all other tasks are done, add analysis tasks. We want
+		// to run them strictly later so that they can leverage the
+		// output of commands executed previously by other tasks, e.g.,
+		// reading resource definitions.
```

Comment:
@philbrookes this is what you asked for. Analysis tasks will run strictly after all others.